clang/lib/Lex/ModuleMap.cpp

```diff
@@ -1307,6 +1307,9 @@ void ModuleMap::setInferredModuleAllowedBy(Module *M,
 
 std::error_code
 ModuleMap::canonicalizeModuleMapPath(SmallVectorImpl<char> &Path) {
+  FileManager &FM = SourceMgr.getFileManager();
+  FM.makeAbsolutePath(Path);
```

Comment:
I'm not 100% sure about this change: I seem to recall some folks are intentionally using relative paths in their invocations in order to be independent of a base path.

Reply:
Are you concerned because this would make the second hash in implicitly-built PCM path depend on CWD? The other place this now gets used is in the `ASTWriter`, but there it gets passed through its `AddPath()`, which makes it relative if needed/requested.

Reply:
This would also impact explicit module invocations.  I didn't realize the impact on path hash; good point that also seems like a concern.  I didn't consider ASTWriter would make it relative, so at least that case wouldn't be an issue.

Reply:
I don't understand how would this impact explicit module invocations? These can still use relative module map paths, it's just that the compiler will use the absolute canonical path internally (IIUC in an non-observable way).

Reply:
Is the idea that because the paths used by the scanner is now indirected through the (implicit) pcm that whatever ASTWriter does to make the path relative will apply to the explicit module invocation?  Before your current changes the scanner was using `canonicalizeModuleMapPath` directly on the path in the invocation, which is what I was naively thinking of.

Reply:
> Is the idea that because the paths used by the scanner is now indirected through the (implicit) pcm that whatever ASTWriter does to make the path relative will apply to the explicit module invocation?

No, because `ASTReader::ReadPath()` does the inverse of `ASTWriter::AddPath()`.

I think we were both talking about different situations. You mean the explicit module invocations generated by the scanner? The canonical module map path would now be absolute. Previously, we'd generate relative paths for module maps found through a relative search paths, or besides includers that themselves were found through a relative path.

I'd be curious about the use-case for this, can you elaborate? Is it being able to reuse the same invocation in directories with different paths, potentially on different machines?

Reply:
> The canonical module map path would now be absolute. Previously, we'd generate relative paths for module maps found through a relative search paths, or besides includers that themselves were found through a relative path.

Yes, this change is what I'm concerned about.

> I'd be curious about the use-case for this, can you elaborate? Is it being able to reuse the same invocation in directories with different paths, potentially on different machines?

My understanding is that this can be used to make the .pcms and command-lines to build them portable across machines, e.g. for caching builds in tools like bazel.  I don't know how much overlap there is between this use case and using the dependency scanner.

Maybe we can flip it around: where exactly do we need an absolute path with your changes?

Reply:
> > The canonical module map path would now be absolute. Previously, we'd generate relative paths for module maps found through a relative search paths, or besides includers that themselves were found through a relative path.
> 
> Yes, this change is what I'm concerned about.

Sorry, I take that back. (It's been a while since I wrote this patch.) This function (and therefore the scanner) used to return absolute paths even before this patch because it always uses `FileManager::getCanonicalName()`.

So the only thing this patch changes is what gets serialized into the PCM file.

When I started canonicalizing the module map path before storing it into the PCM, some tests started unexpectedly failing with `ASTWriter` [reporting inconsistencies](https://github.com/llvm/llvm-project/blob/c60ccfbb898148449946f82cbac6140f1e01cb12/clang/lib/Serialization/ASTReader.cpp#L4093).

When I looked into it, I found it was due to the fact that `FileManager::getCanonicalName()` just forwards the path it's given to `llvm::vfs::FileSystem::getRealPath()`. The problem is, the VFS might have a different idea as to what the CWD is, which breaks with relative module map paths and `-working-directory` (used in `tests/Modules/relative-import-path.c`).

Another issue that popped up was that this function calls `replace_path_prefix(Path, Dir, CanonicalDir)`. If `Dir` is empty, it just slaps the absolute real path for CWD in front of the filename without any separator.

Making the path absolute early on with `FileManager`'s CWD solved both of those issues. (No usage of VFS's CWD, no empty `Dir`.)

So I think given all of that, my concerns are:
* Should we call `FileManager::makeAbsolutePath()` in `ModuleMap::canonicalizeModuleMapPath()` or in `FileManager::getCanonicalName()`?
* Is using `ASTWriter::AddPath()` with the real path problematic? It only performs textual prefix match with a directory path that's not necessarily the real path.

Reply:
> This function (and therefore the scanner) used to return absolute paths even before this patch because it always uses FileManager::getCanonicalName().

Heh, my bad. As the person who wrote that code I should've realized this was already absolute.

> Should we call FileManager::makeAbsolutePath() in ModuleMap::canonicalizeModuleMapPath() or in FileManager::getCanonicalName()?

I think the correct behaviour would be that `FileManager::getCanonicalName` would use `FixupRelativePath` to apply `-working-directory`, but not use `makeAbsolutePath` since that would be inconsistent with how it handles paths in other APIs.  The VFS itself is responsible for resolving CWD in `getRealPath`.

That being said, if this causes other issues I'm fine with your current approach in the meantime.

> Is using ASTWriter::AddPath() with the real path problematic? It only performs textual prefix match with a directory path that's not necessarily the real path.

I don't know how we avoid that without the same issues that motivated canonicalizing in the first place. We don't want to be compiling separate modules for /a/Foo, /a/foo, /a/b/../foo, /symlink/foo, etc.